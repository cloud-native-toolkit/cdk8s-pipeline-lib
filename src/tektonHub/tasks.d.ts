
interface Resource {
  id: number;
  name: string;
}

interface Version {
  id: number;
  version: string;
  rawURL: string;
}

interface Platform {
  id: number;
  name: string;
}
/**
 * This is the expected format of the results from the Tekton API
 * @link https://api.hub.tekton.dev/v1/resources
 */
export interface TektonTask {
  id: number;
  name: string;
  catalog: {
    id: number;
    name: string;
    type: string;
  };
  categories: any[];
  kind: string;
  hubURLPath: string;
  latestVersion: Version;
  tags: any[]; 
  platforms: Platform[];
  rating: number;
}

export interface TektonYaml {
  apiVersion: string = '';
  kind: string = '';
  metadata: {
    name: string;
    labels: { };
    annotations: { };
  };
  spec: {
    description: string;
    workspaces: Workspace[];
    params: Params[];
    results: [ [Object] ];
    steps: [ [Object] ];
  };
}

class Workspace {
  name = '';
  optional: boolean | undefined;
  description = '';
};
class Params {
  name: string = '';
  description: string = '';
  type: string = '';
  default: any;
}