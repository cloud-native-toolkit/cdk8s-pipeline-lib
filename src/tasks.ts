import { PipelineTaskBuilder } from 'cdk8s-pipelines';

export const GitCloneTask = new PipelineTaskBuilder()
  .withName('fetch-source')
  .withTaskReference('git-clone')
  .withWorkspace('output', 'shared-data', 'The files from the git repository')
  .withWorkspace('ssh-directory', 'ssh-creds', 'The SSH keys')
  .withStringParam('url', 'repo-url', '$(params.repo-url)')
;
