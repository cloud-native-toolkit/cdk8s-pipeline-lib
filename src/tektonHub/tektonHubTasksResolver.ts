import { Yaml } from 'cdk8s';
import { ParameterBuilder, Task, TaskBuilder, TaskStepBuilder, WorkspaceBuilder } from 'cdk8s-pipelines';
import { Construct } from 'constructs';

/**
 * This class handles turning a URL that points the YAML for the Tekton Hub Task into a PipelineTask
 */
export class TektonHubTask extends TaskBuilder {
  url: string;
  taskBuild: TaskBuilder;

  /**
   * Creates a new Instance of TektonHubTask with a URL that points to the Raw YAML for the task.
   * @link https://hub.tekton.dev/
   * @param scope
   * @param id
   * @param url string Url to the raw yaml for a Tekton Hub Task (i.e https://raw.githubusercontent.com/tektoncd/catalog/main/task/yq/0.4/yq.yaml)
   */
  constructor(scope: Construct, id: string, url: string) {
    super(scope, id);
    this.url = url;
    this.taskBuild = new TaskBuilder(scope, id);
  }

  private parseYAML(): Boolean {
    const task = this.readYamlFromUrl();
    this.taskBuild.withName(task.metadata?.name!);
    const workspaces = task.spec?.workspaces;
    if (workspaces !== undefined && workspaces?.length !== 0) {
      workspaces.forEach(workspace => {
        this.taskBuild.withWorkspace(new WorkspaceBuilder(workspace.name!)
          .withName(workspace.name!)
          .withDescription(workspace.description!));
      });
    }
    const params = task.spec?.params;
    if (params !== undefined && params.length !== 0) {
      params.forEach(param => {
        this.taskBuild.withStringParam(new ParameterBuilder(param.name!)
          .withDescription(param.description!)
          .withDefaultValue(param.default!)
          .withPiplineParameter(param.name!, param.default!));
      });
    }

    const steps = task.spec?.steps;
    if (steps) {
      steps.forEach(step => {
        const sb =
          new TaskStepBuilder()
            .withName(step.name!)
            .fromScriptData(step.script!)
            .withWorkingDir(step.workingDir!)
            .withArgs(step.args!)
            .withImage(step.image!);
        // step.env?.forEach(e => {
        //   sb.withEnv(e.name);
        // });
        this.taskBuild.withStep(sb);
      });
    }

    return true;
  }

  private readYamlFromUrl(): Task {
    try {
      // Parse the YAML content
      const parsedYaml = Yaml.load(this.url);
      return parsedYaml[0];
    } catch (error) {
      const errorMessage = (error as Error).message; // Type assertion
      throw new Error(`Error reading YAML from URL: ${errorMessage}`);
    }
  }

  /**
   * Returns an instance of PipelineTaskBuilder with the corresponding Tekton Hub Task Link.
   * @returns TaskBuilder
   */
  public build(): TaskBuilder {
    this.parseYAML();
    return this.taskBuild;
  }
}