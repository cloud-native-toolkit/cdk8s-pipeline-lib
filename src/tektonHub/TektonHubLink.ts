import * as fs from 'fs';
import axios from 'axios';
import { TektonTask } from './tasks';

export class TektonHubLink {
  /**
 * Grabs all of the Task defined on Tekton Hub.
 * @link https://hub.tekton.dev
 * @returns Promise<TektonTask[]>
 */
  public async fetchTekonHubTasks(): Promise<TektonTask[]> {
    try {
      const endpoint = 'https://api.hub.tekton.dev/v1/resources';
      const response = await axios.get(endpoint);
      if (response.status !== 200) {
        console.log('Request was not successful.');
        return [];
      }
      const jsonData = response.data;
      if (!jsonData.hasOwnProperty('data')) {
        console.log("'data' property not found in the JSON.");
        return [];
      }
      const data: TektonTask[] = jsonData.data;
      return data;
    } catch (error) {
      throw Error(`Error creating Tekton hub tasks file error: ${error}`);
    }
  }

  private CreateHubTaskFileContent(): Promise<string[]> {
    return this.fetchTekonHubTasks().then((task) => {
      // Make sure we have some tasks
      let fileContents: string[];
      fileContents = [];
      if (task.length === 0) {
        return fileContents;
      }
      // Let's create the tektonHubTask.ts
      // The following code is creating the imports for the Tekton Hub Task
      fileContents.push("import { TektonHubTask } from './tektonHubTasksResolver';");
      fileContents.push("import { Construct } from 'constructs';");
      fileContents.push("import { TaskBuilder } from 'cdk8s-pipelines';");
      task.forEach((item: TektonTask) => {
        if (item.kind !== 'Task') {
          return;
        }
        let name = item.name;
        name = name.replace(/-/g, '_');
        name = name.replace(/[0-9]/g, '');
        name = this.camelize(name);
        const url = item.latestVersion.rawURL;
        // We want to init an instance of tektonHubTaskResolver
        // let str = `export const ${name} = new TektonHubTask('${url}').build();`;
        let str = `export const ${name} = function(scope: Construct, id: string) : TaskBuilder { return new TektonHubTask(scope, id, '${url}').build(); };`;
        fileContents.push(str);
      });
      return fileContents;
    });
  }

  public build() {
    const fileContents = this.CreateHubTaskFileContent();
    const success = fileContents.then((data => {
      try {
        const fp = `${__dirname}/tektonHubTasks.ts`;
        const numberOfTasks = data.length;
        fs.writeFile(fp, data.join('\n'), (err: any) => {
          if (err) {
            throw Error(`Error creating Tekton hub tasks file error: ${err}`);
          }
        });
        console.log(`${numberOfTasks} Tekton Hub task created.`);
        return true;
      } catch (err: any) {
        console.error(err);
        return false;
      }
    }));
    return success;
  }
  /**
   * Helper function to camel case the Task names
   * @param str string to camel case
   * @returns string
   */
  private camelize(str: string): string {
    return str.replace(/(?:^\w|[A-Z]|\b\w)/g, function (word: string, index: number) {
      return index === 0 ? word.toLowerCase() : word.toUpperCase();
    }).replace(/\s+/g, '');
  }
}

