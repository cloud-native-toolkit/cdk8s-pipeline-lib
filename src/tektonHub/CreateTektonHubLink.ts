import { TektonHubLink } from './TektonHubLink';
const tektonHub = new TektonHubLink();
tektonHub.build().then((success => {
  if (!success) {
    throw Error('Unable to build Tekton Hub Link');
  }
  console.log('Tekton hub task completed!');
})).catch(reason => {
  console.error(reason);
});