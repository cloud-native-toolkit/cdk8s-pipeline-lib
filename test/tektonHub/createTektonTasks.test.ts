import fs from 'fs';
import axios from 'axios';
// import { Testing } from 'cdk8s';
import fetchData from './../__mocks__/tektonHub.json';
import { TektonTask } from '../../src/tektonHub/tasks';
import { TektonHubLink } from '../../src/tektonHub/TektonHubLink';

jest.mock('axios');
jest.mock('fs');

describe('testTekonFetch', () => {
  test('tektonHub', async () => {
    (axios.get as jest.MockedFunction<typeof axios.get>).mockResolvedValue({
      status: 200,
      data: fetchData,
    });
    const tetkontHubLink = new TektonHubLink();
    const tektonHubTasks = await tetkontHubLink.fetchTekonHubTasks();
    expect(tektonHubTasks).toEqual(expect.any(Array<TektonTask>)); // Check if the result is an array of TektonTask
    expect(tektonHubTasks).toHaveLength(6); // Check if the array length matches
  });

  test('tektonHubInvalidData', async () => {
    (axios.get as jest.MockedFunction<typeof axios.get>).mockResolvedValue({
      status: 500,
      data: [],
    });
    try {
      const logSpy = jest.spyOn(console, 'log');
      const tetkontHubLink = new TektonHubLink();
      const result = await tetkontHubLink.fetchTekonHubTasks();
      expect(logSpy).toHaveBeenCalledWith('Request was not successful.');
      expect(result).toBe([]);
    } catch (Error: any) {
      return;
    }
  });

  test('testFileCreationData', async () => {
    const logSpy = jest.spyOn(console, 'log');
    expect(jest.isMockFunction(fs.writeFile)).toBeTruthy();
    (axios.get as jest.MockedFunction<typeof axios.get>).mockResolvedValue({
      status: 200,
      data: fetchData,
    });
    const tetkontHubLink = new TektonHubLink();
    // @ts-ignore: We temporarily ignore TypeScript's error
    await tetkontHubLink.CreateHubTaskFileContent().then(data => {
      expect(data).toMatchSnapshot();
    });
    const tektonBuildFile = await tetkontHubLink.build();
    expect(logSpy).toHaveBeenCalledWith('8 Tekton Hub task created.');
    expect(tektonBuildFile).toBeTruthy;
  });

});
