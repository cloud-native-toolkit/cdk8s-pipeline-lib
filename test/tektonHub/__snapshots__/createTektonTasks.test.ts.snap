// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testTekonFetch testFileCreationData 1`] = `
[
  "import { TektonHubTask } from './tektonHubTasksResolver';",
  "import { Construct } from 'constructs';",
  "import { TaskBuilder } from 'cdk8s-pipelines';",
  "export const git_cli = function(scope: Construct, id: string) : TaskBuilder { return new TektonHubTask(scope, id, 'https://raw.githubusercontent.com/tektoncd/catalog/main/task/git-cli/0.4/git-cli.yaml').build(); };",
  "export const gogit = function(scope: Construct, id: string) : TaskBuilder { return new TektonHubTask(scope, id, 'https://raw.githubusercontent.com/tektoncd/catalog/main/task/gogit/0.1/gogit.yaml').build(); };",
  "export const golang_build = function(scope: Construct, id: string) : TaskBuilder { return new TektonHubTask(scope, id, 'https://raw.githubusercontent.com/tektoncd/catalog/main/task/golang-build/0.3/golang-build.yaml').build(); };",
  "export const jenkins = function(scope: Construct, id: string) : TaskBuilder { return new TektonHubTask(scope, id, 'https://raw.githubusercontent.com/tektoncd/catalog/main/task/jenkins/0.1/jenkins.yaml').build(); };",
  "export const pull_request = function(scope: Construct, id: string) : TaskBuilder { return new TektonHubTask(scope, id, 'https://raw.githubusercontent.com/tektoncd/catalog/main/task/pull-request/0.1/pull-request.yaml').build(); };",
]
`;
